File: seastarvideo/src/main/java/com/freewind/seastarvideo/api/ApiEngine.java
package com.freewind.seastarvideo.api;

import androidx.annotation.NonNull;
import com.freewind.seastarvideo.BuildConfig;
import com.freewind.seastarvideo.bean.BaseBean;
import com.google.gson.Gson;
import com.shiyuan.meeting.api.HttpLogCollectorInterceptor;
import com.shiyuan.meeting.util.EncryptUtil;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.TimeUnit;
import okhttp3.Cache;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.logging.HttpLoggingInterceptor;
import okio.Buffer;
import retrofit2.Call;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/**
 * author superK
 * update_at 2024/1/29
 * description 网络请求引擎
 */
public class ApiEngine {

    public static final String BaseUrl = "http://localv2.srtc.live:8089/api/";
    private static final String METHOD_GET = "GET";
    private static final String METHOD_POST = "POST";
    private OkHttpClient okHttpClient;
    private Retrofit retrofit;
    private int connectTimeout = 30,readTimeout = 30,writeTimeout = 30;
    private boolean reconnect = true;

    public OkHttpClient getOkHttpClient() {
        return okHttpClient;
    }

    public Retrofit getRetrofit() {
        return retrofit;
    }
    Gson gson = new Gson();



    /**
     * 初始化引擎
     * @param baseUrl 请求前缀
     * @param appId 应用 id
     * @param appKey 应用 key
     */
    public void init(String baseUrl, String appId, String appKey){
        if (baseUrl == null){
            return;
        }
        if (!baseUrl.endsWith("/")){
            baseUrl = baseUrl + "/";
        }
        release();
        if (okHttpClient == null){
            OkHttpClient.Builder okhttpBuilder = new OkHttpClient.Builder()
                    .connectTimeout(connectTimeout, TimeUnit.SECONDS)//设置连接超时
                    .readTimeout(readTimeout, TimeUnit.SECONDS)//设置读超时
                    .writeTimeout(writeTimeout,TimeUnit.SECONDS)//设置写超时
                    .retryOnConnectionFailure(reconnect);//是否自动重连
            okhttpBuilder.addInterceptor(chain -> {
                Request original = chain.request();
                StringBuilder stringBuilder = new StringBuilder();

                String nonce = UUID.randomUUID().toString().replace("-", "").substring(0,16);
                long timestamp = System.currentTimeMillis()/1000;
                String signature = EncryptUtil.hmacSHA256Encrypt(stringBuilder.toString(), appKey);
                List<String> names = new ArrayList<>();
                names.add("app_id=" + appId);
                names.add("nonce=" + nonce);
                names.add("timestamp=" + timestamp);
                names.add("signature=" + signature);
                if (METHOD_GET.equals(original.method())) {
                    Set<String> paramKeys = original.url().queryParameterNames();
                    for (String key : paramKeys) {
                        names.add(key + "=" + original.url().queryParameter(key));
                    }
                }
                Collections.sort(names, String::compareTo);
                for (String name : names) {
                    stringBuilder.append(name).append("&");
                }

                String json = "";
                if (METHOD_GET.equals(original.method())) {
                    stringBuilder.deleteCharAt(stringBuilder.length() - 1);
                }else if (METHOD_POST.equals(original.method())){
                    RequestBody body = original.body();
                    if (body != null){
                        json = getParamContent(body);
                        stringBuilder.append(json);
                    }
                }

                Request request = original.newBuilder()
                        .header("app_id", appId)
                        .header("nonce", nonce)
                        .header("timestamp", timestamp+"")
                        .header("signature", signature)
//                        .method(original.method(), original.body())
                        .post(RequestBody.create(json, MediaType.get("application/json")))
                        .build();

                return chain.proceed(request);
            });
            okhttpBuilder.addInterceptor(new HttpLogCollectorInterceptor());
            if (BuildConfig.DEBUG) {
                HttpLoggingInterceptor logging = new HttpLoggingInterceptor();
                logging.setLevel(HttpLoggingInterceptor.Level.BODY);
                okhttpBuilder.addInterceptor(logging);
            }
            okHttpClient = okhttpBuilder.build();
        }
        if (retrofit == null){
            retrofit = new Retrofit.Builder()
                    .client(okHttpClient)
                    .baseUrl(baseUrl)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }
    }

    /**
     * 获取常规post请求参数
     */
    private String getParamContent(RequestBody body) throws IOException {
        Buffer buffer = new Buffer();
        body.writeTo(buffer);
        String json = buffer.readUtf8();
        json = json.replace("\"{", "{");
        json = json.replace("}\"", "}");
        return json.replace("\\\"", "\"");
    }

    public <T> T create(final Class<T> service){
        if (retrofit != null){
            return retrofit.create(service);
        }
        return null;
    }

    /**
     * 异步请求队列
     * @param call  请求
     * @param apiCallback 回调
     * @param <T> 类型
     */
    public <T extends BaseBean> void enqueue(Call<T> call, Callback<T> apiCallback){
        if (apiCallback != null){
            apiCallback.onStart();
        }
        call.enqueue(new retrofit2.Callback<T>() {
            @Override
            public void onResponse(@NonNull Call<T> call, @NonNull Response<T> response) {
                T t = response.body();
                if (t != null){
                    if (t.getCode() == ApiCode.SUCCESS){
                        if (apiCallback != null){
                            apiCallback.onSuccess(t);
                        }
                    }else {
                        if (apiCallback != null){
                            apiCallback.onFailed(t.getCode(), t.getMsg());
                        }
                    }
                }else {
                    if (apiCallback != null){
                        apiCallback.onFailed(ApiCode.ERROR, "body为空");
                    }
                }
            }

            @Override
            public void onFailure(@NonNull Call<T> call, @NonNull Throwable throwable) {
                if (apiCallback != null){
                    if (call.isCanceled()){
                        apiCallback.onCancel();
                        return;
                    }
                    apiCallback.onFailed(ApiCode.ERROR, throwable.getMessage());
                }
            }
        });
    }

    /**
     * 取消队列中的请求和进行中的请求
     * @param requestKey    key
     */
    public void cancelCallWithKey(String requestKey){
        if (okHttpClient == null){
            return;
        }
        if (requestKey == null || requestKey.isEmpty()){
            return;
        }
        //When you want to cancel:
        //A) go through the queued calls and cancel if the tag matches:
        for (okhttp3.Call call : okHttpClient.dispatcher().queuedCalls()) {
            String tag = (String) call.request().tag();
            if (tag != null && tag.startsWith(requestKey)){
                call.cancel();
                break;
            }
        }

        //B) go through the running calls and cancel if the tag matches:
        for (okhttp3.Call call : okHttpClient.dispatcher().runningCalls()) {
            String tag = (String) call.request().tag();
            if (tag != null && tag.startsWith(requestKey)){
                call.cancel();
                break;
            }
        }
    }

    /**
     * 取消队列中的请求和进行中的请求
     */
    public void cancelCallForAll(){
        if (okHttpClient == null){
            return;
        }
        //When you want to cancel:
        //A) go through the queued calls and cancel if the tag matches:
        for (okhttp3.Call call : okHttpClient.dispatcher().queuedCalls()) {
            call.cancel();
        }

        //B) go through the running calls and cancel if the tag matches:
        for (okhttp3.Call call : okHttpClient.dispatcher().runningCalls()) {
            call.cancel();
        }
    }

    /**
     * 释放资源
     */
    public void release(){
        cancelCallForAll();
        if (okHttpClient != null){
//            okHttpClient.interceptors().clear();//清除拦截器
            okHttpClient.dispatcher().executorService().shutdown();   //清除并关闭线程池
            okHttpClient.connectionPool().evictAll();                 //清除并关闭连接池
            Cache cache = okHttpClient.cache();
            try {
                if (cache != null){
                    cache.close();                             //清除cache
                }
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }
        retrofit = null;
        okHttpClient = null;
    }

}

File: seastarvideo/src/main/java/com/freewind/seastarvideo/http/ApiEngine.kt
package com.freewind.seastarvideo.http

import com.freewind.seastarvideo.BuildConfig
import com.freewind.seastarvideo.EnvArgument
import com.freewind.seastarvideo.http.bean.BaseBean
import com.freewind.seastarvideo.http.interceptor.HeaderInterceptor
import com.freewind.seastarvideo.http.interceptor.ResponseInterceptor
import okhttp3.OkHttpClient
import okhttp3.RequestBody
import okhttp3.logging.HttpLoggingInterceptor
import okio.Buffer
import retrofit2.Call
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * author superK
 * update_at 2024/1/29
 * description 网络请求引擎
 */
class ApiEngine {
    companion object {
        const val BaseUrl = "http://localv2.srtc.live:8089/api/"
    }
    var okHttpClient: OkHttpClient? = null
        private set
    var retrofit: Retrofit? = null
        private set
    // 连接超时时间
    private val connectTimeout = 30
    // 读取超时时间
    private val readTimeout = 30
    // 写入超时时间
    private val writeTimeout = 30
    // 是否自动重连
    private val reconnect = true

    /**
     * 初始化引擎
     * @param baseUrl 请求前缀
     */
    fun init(baseUrl: String) {
        if (baseUrl.isEmpty()) {
            return
        }
        var url = baseUrl
        if (!url.endsWith("/")) {
            url = "$url/"
        }
        release()
        if (okHttpClient == null) {
            val okhttpBuilder: OkHttpClient.Builder = OkHttpClient.Builder()
                .connectTimeout(connectTimeout.toLong(), TimeUnit.SECONDS) //设置连接超时
                .readTimeout(readTimeout.toLong(), TimeUnit.SECONDS) //设置读超时
                .writeTimeout(writeTimeout.toLong(), TimeUnit.SECONDS) //设置写超时
                .retryOnConnectionFailure(reconnect) //是否自动重连
            okhttpBuilder.addInterceptor(HeaderInterceptor())
            okhttpBuilder.addInterceptor(ResponseInterceptor())
//            okhttpBuilder.addInterceptor(HttpLogCollectorInterceptor())
            if (BuildConfig.DEBUG) {
                val logging = HttpLoggingInterceptor()
                logging.setLevel(HttpLoggingInterceptor.Level.BODY)
                okhttpBuilder.addInterceptor(logging)
            }

            okHttpClient = okhttpBuilder.build()
        }
        if (retrofit == null) {
            retrofit = Retrofit.Builder()
                .client(okHttpClient!!)
                .baseUrl(url)
                .addConverterFactory(GsonConverterFactory.create())
                .build()
        }
    }

    /**
     * 获取常规post请求参数
     */
    @Throws(IOException::class)
    private fun getParamContent(body: RequestBody): String {
        val buffer = Buffer()
        body.writeTo(buffer)
        var json = buffer.readUtf8()
        json = json.replace("\"{", "{")
        json = json.replace("}\"", "}")
        return json.replace("\\\"", "\"")
    }

    fun <T> create(service: Class<T>): T? {
        return if (retrofit != null) {
            retrofit!!.create(service)
        } else {
            null
        }
    }

    /**
     * 异步请求队列
     * @param call  请求
     * @param apiCallback 回调
     * @param <T> 类型
    </T> */
    fun <T : BaseBean> enqueue(call: Call<T>, apiCallback: Callback<T>?) {
        apiCallback?.onStart()
        call.enqueue(object : retrofit2.Callback<T> {
            override fun onResponse(call: Call<T>, response: Response<T>) {
                val t = response.body()
                if (t != null) {
                    if (t.code == ApiCode.SUCCESS) {
                        apiCallback?.onSuccess(t)
                    } else if (t.code == ApiCode.ERROR_AUTH_FAILED ||
                        t.code == ApiCode.ERROR_TOKEN_INVALID ||
                        t.code == ApiCode.ERROR_TOKEN_TOKEN_EXPIRED) {
                        // 如果错误是“未登录”、“token失效”、“token已过期”，则需要跳转到登录页面重新做登录操作
                        EnvArgument.instance.goToLoginPage()
                        apiCallback?.onFailed(t.code, t.msg)
                    } else {
                        apiCallback?.onFailed(t.code, t.msg)
                    }
                } else {
                    apiCallback?.onFailed(ApiCode.ERROR, "body为空")
                }
            }

            override fun onFailure(call: Call<T>, throwable: Throwable) {
                if (apiCallback != null) {
                    if (call.isCanceled) {
                        apiCallback.onCancel()
                        return
                    }
                    apiCallback.onFailed(ApiCode.ERROR, throwable.message)
                }
            }
        })
    }

    /**
     * 取消队列中的请求和进行中的请求
     * @param requestKey    key
     */
    fun cancelCallWithKey(requestKey: String) {
        if (okHttpClient == null) {
            return
        }
        if (requestKey.isEmpty()) {
            return
        }
        //When you want to cancel:
        //A) go through the queued calls and cancel if the tag matches:
        for (call in okHttpClient!!.dispatcher.queuedCalls()) {
            val tag = call.request().tag() as String?
            if (tag != null && tag.startsWith(requestKey)) {
                call.cancel()
                break
            }
        }

        //B) go through the running calls and cancel if the tag matches:
        for (call in okHttpClient!!.dispatcher.runningCalls()) {
            val tag = call.request().tag() as String?
            if (tag != null && tag.startsWith(requestKey)) {
                call.cancel()
                break
            }
        }
    }

    /**
     * 取消队列中的请求和进行中的请求
     */
    fun cancelCallForAll() {
        if (okHttpClient == null) {
            return
        }
        //When you want to cancel:
        //A) go through the queued calls and cancel if the tag matches:
        for (call in okHttpClient!!.dispatcher.queuedCalls()) {
            call.cancel()
        }

        //B) go through the running calls and cancel if the tag matches:
        for (call in okHttpClient!!.dispatcher.runningCalls()) {
            call.cancel()
        }
    }

    /**
     * 释放资源
     */
    fun release() {
        cancelCallForAll()
        if (okHttpClient != null) {
//            okHttpClient.interceptors().clear();//清除拦截器
            okHttpClient!!.dispatcher.executorService.shutdown() //清除并关闭线程池
            okHttpClient!!.connectionPool.evictAll() //清除并关闭连接池
            val cache = okHttpClient!!.cache
            try {
                cache?.close()
            } catch (e: IOException) {
                throw RuntimeException(e)
            }
        }
        retrofit = null
        okHttpClient = null
    }
}
